### ISSUE TITLE
support Go 1.15

### ISSUE THREAD

**Issue by mvdan** [lucas-clemente/quic-go#2614]
As of https://go-review.googlesource.com/c/go/+/234237 last week, tls.ClientSessionState has been modified, so I see panics on the latest version:

```
panic: qtls.ClientSessionState not compatible with tls.ClientSessionState

goroutine 1 [running]:
github.com/lucas-clemente/quic-go/internal/handshake.init.2()
	/home/mvdan/go/pkg/mod/github.com/lucas-clemente/quic-go@v0.17.1/internal/handshake/unsafe.go:26 +0x205
```

I assume this requires a bit of extra code to handle the go1.15 build tag. beta1 was released recently, and thankfully it includes that commit, so you can add that tagged version to CI to test the fix. RC1 should be out in about two weeks, and I doubt they'll change the tls package further.

**Comment by marten-seemann** [comment]
We probably won't add support for Go 1.15 before the final release. We currently need to support both Go 1.13 and Go 1.14, since libp2p / IFPS are still using Go 1.13. This already leads to significant complexity in both qtls and quic-go. Supporting Go 1.15 at the time would make things even more complicated.

**Comment by mvdan** [comment]
Thanks, I understand the added cost of supporting multiple Go versions. I personally support two stable versions at a time for my libraries, so right now that is 1.13 and 1.14 as well.

However, I disagree that we should wait until the final Go 1.15 release. We all need to try and test with Go 1.15 pre-releases (beta1, rc1, rc2) before the final release is out. If the final release has a big regression or design mistake, it's too late - it will take at least another couple of weeks for a fix to be released, if we can even fix the problem without breaking backwards compatibility.

It is practically impossible to test Go 1.15 at all with any library that indirectly imports quic-go, because quic-go will panic at init time. I don't even know if I use QUIC, but since it's built and linked into my binaries, the effect is the same.

If you don't want to support both Go 1.13 and 1.15 at the same time, that's understandable. I would suggest that you move master to support 1.14 and 1.15, making sure that the latest stable release supports 1.13 until the final 1.15 release. That way, none of the existing users can break, and I can use master with 1.15 until the final release triggers a proper quic-go release.

**Comment by mvdan** [comment]
Also, I would totally understand "sorry but I don't have time right now, PRs are welcome, otherwise I'll get to it once 1.15 is out". That's not what I gathered from your message, though.

**Comment by mvdan** [comment]
With 1.13 support dropped a month ago, and go1.15rc1 having come out earlier this week, would you accept a patch for this now? Any tips? I'm still blocked on testing my project with 1.15 because of the panic at init time.

**Comment by cryptix** [comment]
@mvdan looking at [where the panic happens](https://github.com/lucas-clemente/quic-go/blob/master/internal/handshake/unsafe.go#L20), I think you need to patch `github.com/marten-seemann/qtls`. I tried but sadly the process for updating it / the  changes on top of `crypto/tls` are not that obvious.

**Comment by marten-seemann** [comment]
That's right, the qtls upgrade is the tricky part here. If we wanted to support 1.14 and 1.15 at the same time, we'd have to create two copies of every Go file in the qtls repo and make sure that each of the two copies has the appropriate build flags such that it only builds on one of those versions.

**Comment by mvdan** [comment]
Hm. If that's too much trouble, I assume an alternative would be to keep Go 1.14 support in v0.17, and release v0.18 with just Go 1.15 support. Then backport any important bugfixes to v0.17 if necessary. Not ideal, but having the latest version only work with Go 1.15 is better than having it only work with 1.14.

**Comment by vext01** [comment]
This has also broken the Syncthing and go-ipfs binary packages on OpenBSD (and probably Linux distros too). We'd be super-grateful for a patch to fix this.

**Comment by AudriusButkevicius** [comment]
I'm piling in on the winging here from Syncthing's perspective. 

We want to move to Go 1.15 as soon as we physically can, and having to drop quic support in order to do that would be grim.

**Comment by marten-seemann** [comment]
Hi guys, my apologies for this mess. This clearly hasn't been a good way to deal with a Go version upgrade.
I'm working on the Go 1.15 update right now and I expect to ship an updated version of quic-go before / on the weekend.

It looks like it actually is possible to use multiple versions of same Go module: https://www.percona.com/blog/2020/03/09/using-different-versions-of-a-package-in-an-application-via-go-modules/. I just found out about this trick today. This will alleviate a lot of the pains that come with supporting multiple Go versions at the same time:
We could import `v.x` of qtls for Go 1.14x and `v.y` for Go 1.15.x, so we don't need do anything ugly with build flags in qtls. We'll probably just need two copies of [unsafe.go](https://github.com/lucas-clemente/quic-go/blob/master/internal/handshake/unsafe.go) and [qtls.go](https://github.com/lucas-clemente/quic-go/blob/master/internal/handshake/qtls.go), one for 1.14 and one for 1.15.

**Comment by imsodin** [comment]
Thanks for looking into this!

Replace directives unfortunately don't work here:

> replace directives only apply in the main module's go.mod file and are ignored in other modules. See Minimal version selection for details.  
> https://golang.org/ref/mod#go

So anyone using quic as a library won't have that benefit (probably everyone, not sure whether you produce any binaries/have a main package?). Build tags feels like the best way to go. Alternatively what @mvdan proposed: Keep <=go1.14 compat in 0.17.x (probably only doing bugfixes there) and >=1.15 in 0.18.x.

**Comment by mvdan** [comment]
Just for the sake of suggesting alternatives, if you don't want `go get -u` to suddenly break support for 1.14, you could also use a separate branch for 1.15 and tell people to `go get github.com/lucas-clemente/quic-go@branchname`. The downside being that they'd have to do that explicitly.

**Comment by marten-seemann** [comment]
> So anyone using quic as a library won't have that benefit (probably everyone, not sure whether you produce any binaries/have a main package?). Build tags feels like the best way to go. Alternatively what @mvdan proposed: Keep <=go1.14 compat in 0.17.x (probably only doing bugfixes there) and >=1.15 in 0.18.x.

What a bummer. So if importing multiple versions of the same module doesn't work, I could split up qtls into two repositories, one for Go 1.14 and one for Go 1.15. Build flags in quic-go would then make sure that we use the right repo depending on the Go version.

**Comment by Lekensteyn** [comment]
The ClientSessionState type gained two new private fields in https://github.com/golang/go/commit/7b872b6d955d3e749ea62dbfced68ab5c61eae91

It should be possible to move just that single `type ClientSession struct { ... }` field in a separate `defs.go` and `defs114.go` file with annotations `// +build !go1.14` and `// +build go1.14` respectively. Repeat for other structures, but as far as I can see this is the only incompatible structure. This approach should be better than duplicating the whole common.go file which would result in a lot of implementation duplication.

**Comment by marten-seemann** [comment]
There were a lot more changes to `crypto/tls` in Go 1.15. I can't really afford to add build flags for every change between the two versions, so having a clean commit history based off Go 1.14 and Go 1.15 seems like the easier solution here, even if it leads to code duplication.

**Comment by marten-seemann** [comment]
Quick update:
I've created a qtls that works with Go 1.15. Unfortunately, the changes they made to `crypto/tls` break session resumption / 0-RTT in QUIC if `Config.SessionTicketKeys` (which is now deprecated) is not set: In Go 1.14, `Config.Clone()` would generate [a session ticket key before actually cloning the config](https://github.com/golang/go/blob/d571a77846dfee8efd076223a882915cd6cb52f4/src/crypto/tls/common.go#L661-L664).
In Go 1.15, the session ticket key is generated lazily [when the it is used for the first time](https://github.com/golang/go/blob/0fdc3801bfd43d6f55e4ea5bf095e1ea55430339/src/crypto/tls/common.go#L810-L863). That means that if a QUIC server is started with a `tls.Config` that doesn't set the `SessionTicketKeys`, it will generate new session ticket keys for every QUIC connection (as we convert the `tls.Config` to a `qtls.Config` when accepting a new connection).
A possible solution for this would be to only convert the `tls.Config` to `qtls.Config` once for every `quic.Listen` call. The problem with this approach is twofold:
1. If the same `tls.Config` is used for multiple `quic.Listen` calls, they will end up with different session ticket keys. This might not be too bad in practice, but it's a difference from using the same `tls.Config` for two `tls.Listen` calls.
2. As `SessionTicketKeys` is deprecated now, it looks like we're now supposed to use `Config.SetSessionTicketKeys`. As soon as we have converted a `tls.Config` to a `qtls.Config`, calls to `tls.Config.SetSessionTicketKeys` won't have any effect any more, effectively disabling key rotation. This is bad.

My current preference goes towards restoring the Go 1.14 behavior of generating session ticket keys *before* they're actually used. All that's needed for that is calling `tls.Config.ticketKeys(nil)`. Of course, this will require yet another use of `unsafe`, but that ship has sailed a long time ago anyway.

**Comment by swills** [comment]
The update to Go 1.15 also broke this and Syncthing (and presumably ipfs-go) in FreeBSD ports too.

**Comment by mholt** [comment]
Subscribing, since this is blocking the Caddy 2.2 release.

**Comment by vext01** [comment]
Sorry to be persistent, but we are getting quite desperate for a fix for Syncthing.

Is there some kind of minimal patch that could be applied to make quic work *only* with go-1.15? That'd get us going.

**Comment by marten-seemann** [comment]
#2720 is the fix. Just waiting for code review, then I can merge and cut a new release.

**Comment by NatoBoram** [comment]
> @lucas-clemente lucas-clemente approved these changes 

Looks like said code review is already done 🎉

**Comment by ranisalt** [comment]
Seriously, a *runtime* deploying backward-incompatible changes in a **MINOR** version update? It had to be Go...

**Comment by swills** [comment]
> Seriously, a _runtime_ deploying backward-incompatible changes in a **MINOR** version update? It had to be Go...

That's a general Go issue and not specific to this project or issue, so not really relevant here. Please don't distract those of us monitoring this issue for resolution with irrelevant topics.

**Comment by mholt** [comment]
@ranisalt 

> Seriously, a runtime deploying backward-incompatible changes in a MINOR version update? It had to be Go...

The change wasn't in the runtime, it was in the standard library. And use of [the `unsafe` package is not covered under the compatibility promise](https://golang.org/pkg/unsafe/):

> Packages that import unsafe may be non-portable and are not protected by the Go 1 compatibility guidelines.

Let's keep this on track. Clearly, there is a lot of complexity being hidden from us and we're lucky to be able to use a complete HTTP/3 package so that we don't have to do this ourselves.

**Comment by marten-seemann** [comment]
Hi guys, I just released [v0.18.0](https://github.com/lucas-clemente/quic-go/releases/tag/v0.18.0), which adds support for Go 1.15.

**Sorry for all the trouble caused!**
This was a very suboptimal way to handle a Go version update, and we'll make sure that the Go 1.16 will go more smoothly. Ideally, this means supporting beta / release candidate versions, as @mvdan suggested in the original issue.

**Comment by mvdan** [comment]
Thank you @marten-seemann - the speed and future improvement is really appreciated. I realise that pressure on open source developers can be a lot, and I'm happy to help in the future in any way I can. I regularly use the master version of Go, so at least I can quickly flag when master breaks again in the future.

**Comment by marten-seemann** [comment]
@mvdan That would be very much appreciated! The new design with one qtls repository per Go version, while super ugly, will allow us to easily add support for another Go version with minimal overhead, so I'd be happy to add a new one for Go 1.16 once master has accumulated non-backwards-compatible changes.

**Comment by mholt** [comment]
Thanks so much for your hard work on this @marten-seemann ! Looking forward to tagging Caddy 2.2 soon :)
